src/aleph/handlers/forget.py

```diff
@@ -131,11 +131,31 @@ async def forget_if_allowed(target_message: Dict, forget_message: ForgetMessage)
         await garbage_collect(store_content.item_hash, store_content.item_type)
 
 
-async def handle_forget_message(message: Dict, content: Dict):
+async def handle_forget_message(message: Dict, content: Dict) -> Optional[bool]:
+    """
+    Processes a FORGET message by deleting the content of all targeted messages.
+
+    :param message: The FORGET message, as a dictionary.
+    :param content: The JSON item_content field of the message, as a dictionary.
+    :return: True if the FORGET message was processed correctly. None if at least
+             one of the hashes is currently missing on the node, meaning that
+             the message will be marked for retry at a later time.
+    """
+
     # Parsing and validation
     forget_message = ForgetMessage(**message, content=content)
     logger.debug(f"Handling forget message {forget_message.item_hash}")
 
+    hashes_to_forget = forget_message.content.hashes
+    matching_messages = Message.collection.find(
+        {"item_hash": {"$in": forget_message.content.hashes}}, {"item_hash": 1}
+    )
+    matching_hashes = set([msg["item_hash"] async for msg in matching_messages])
+
+    if matching_hashes != set(hashes_to_forget):
+        # Mark the FORGET message for retry
+        return None
```

Comment:
Yes that's a side-effect of the proposed implementation. Anyway, we have a problem with pending messages that never get processed. I was thinking that we should add a maximum amount of retries, or a timeout (ex: a message that has not been processed in 72 hours should be dropped). But this implies the need to better control the versions of the CCN across the network, as retries also happen in case of a bug on our side.

Also, there's a race condition that can occur with what you propose (when the message is still in the pipeline), with this implementation we guarantee that the message will be forgotten. But overall I agree that this mechanism could be improved, I just haven't had a better idea yet.

Reply:
An option for this PR is to check the message time and drop it if it's been retried a few times and is older than 72 hours.

Reply:
I believe that when a FORGET messages references multiple hashes, all valid hashes should be processed. The message may be flagged to be retried later as well if all hashes are not found, but existing hashes should be forgotten as soon as possible.

Reply:
If the processing was synchronous and we could throw an error (HTTP 400 Bad Request) at the user directly, I would do that. But since the processing is asynchronous after user acknowledgement, I believe we should do a best effort to process user content.